geps/gep-1494/index.md

```diff
@@ -130,20 +130,383 @@ From @ongy, some additional goals to keep in mind:
+
+* something like ext_auth was a good idea, because it's flexible and allows the
+  implementation of many types of Auth without specific protocol implementation
+  in upstream
+* Envoy's ext_auth protocol has no major problems that would stop us using it
+* Envoy-based implementations mostly already have support for it
+
+At that meeting, those present agreed that ext_auth was a good place to start.
+
+Most non-Envoy dataplanes also already have similar methods, so the maintainers
+of projects using other dataplanes were okay with this idea.
+
+The alternative here would be to add a Filter type _per auth method_, which, given
+the large number of options, could quickly become very complex.
+
+This GEP is, however, explicitly _not_ ruling out the possibility of adding
+specific Filters for specific Auth methods in the future, if users of this API
+find the overhead of running a compatible implementation to be too much.
+
+### API Design
+
+#### Phase 1: Adding a Filter
+
+This config mainly takes inspiration from Envoy's ext_auth filter config, while
+also trying to maintain compatibility with other HTTP methods.
+
+This design is also trying to start with a minimum feature set, and add things
+as required, rather than adding everything configurable in all implementations
+immediately.
+
+There is some difference between data planes, based on the links above, but
+these fields should be broadly supportable across all the listed implementations.
+
+Some design comments are included inline.
+
+The intent for Phase 2 is that this struct will be included in an eventual Policy
+so that additions only need to be made in one place.
+
+Additionally, as per other added Filters, the config is included in HTTPRoute,
+and is not an additional CRD.
+
+##### Go Structs
+
+```go
+
+// HTTPRouteExtAuthProtcol specifies what protocol should be used
+// for communicating with an external authorization server.
+//
+// Valid values are supplied as constants below.
+type HTTPRouteExtAuthProtocol string
+
+const (
+	HTTPRouteExtAuthGRPCProtocol HTTPRouteExtAuthProtocol = "GRPC"
+	HTTPRouteExtAuthHTTPProtocol HTTPRouteExtAuthProtocol = "HTTP"
+)
+// HTTPExtAuthFilter defines a filter that modifies requests by sending
+// request details to an external authorization server.
+//
+// Support: Extended
+// Feature Name: HTTPRouteExtAuth
+type HTTPExtAuthFilter struct {
+
+	// ExtAuthProtocol describes which protocol to use when communicating with an
+	// ext_auth authorization server.
+	//
+	// When this is set to GRPC, each backend must use the Envoy ext_auth protocol
+	// on the port specified in `backendRefs`. Requests and responses are defined
+	// in the protobufs explained at:
+	// https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/external_auth.proto
+	//
+	// When this is set to HTTP, each backend must respond with a status code in
+	// the `2xx` range on a successful authorization. Any other code is considered
+	// an authorization failure.
```

Comment:
This is actually not quite correct. `200` and _only_ `200` means "successfully authorized"; _anything_ else, including any other `2xx` response, must be sent back to the client.

(I realize that anything other than `200` from the extauth server would be very rare, but that's the way the HTTP variant was designed.)

Reply:
Where is this defined?

Reply:
😐 Officially, formally... nowhere. (I mean, there's draft on my Mac, but that doesn't count.)

Aside from reading Envoy source, here's the only written [doc for ext_authz] I've ever been able to find describing the protocol. That dates back to just after we created the protocol, and includes the note about `200`, but it's obviously not a real spec.

[doc for ext_authz]: https://www.getambassador.io/docs/emissary/latest/topics/running/services/ext-authz.

Reply:
@howardjohn Well, the link now doesn't 404, since I've fixed it to include the ".html" on the end. 😅 

BUT you're absolutely right. Hoping to get that doc off my MacBook shortly. 🙂

Reply:
Okay, I'll change to `200`, and then we can document the protocol in an appendix or something in a future PR to this GEP.